// @aws-cdk/aws-elasticache/lib/elasticache-canned-metrics.generated.js
"use strict";
// Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElastiCacheMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class ElastiCacheMetrics {
    static cpuUtilizationAverage(dimensions) {
        return {
            namespace: 'AWS/ElastiCache',
            metricName: 'CPUUtilization',
            dimensions,
            statistic: 'Average',
        };
    }
    static freeableMemoryAverage(dimensions) {
        return {
            namespace: 'AWS/ElastiCache',
            metricName: 'FreeableMemory',
            dimensions,
            statistic: 'Average',
        };
    }
    static swapUsageAverage(dimensions) {
        return {
            namespace: 'AWS/ElastiCache',
            metricName: 'SwapUsage',
            dimensions,
            statistic: 'Average',
        };
    }
    static networkBytesInAverage(dimensions) {
        return {
            namespace: 'AWS/ElastiCache',
            metricName: 'NetworkBytesIn',
            dimensions,
            statistic: 'Average',
        };
    }
    static networkBytesOutAverage(dimensions) {
        return {
            namespace: 'AWS/ElastiCache',
            metricName: 'NetworkBytesOut',
            dimensions,
            statistic: 'Average',
        };
    }
    static currConnectionsSum(dimensions) {
        return {
            namespace: 'AWS/ElastiCache',
            metricName: 'CurrConnections',
            dimensions,
            statistic: 'Sum',
        };
    }
    static evictionsSum(dimensions) {
        return {
            namespace: 'AWS/ElastiCache',
            metricName: 'Evictions',
            dimensions,
            statistic: 'Sum',
        };
    }
    static reclaimedSum(dimensions) {
        return {
            namespace: 'AWS/ElastiCache',
            metricName: 'Reclaimed',
            dimensions,
            statistic: 'Sum',
        };
    }
    static cacheHitsSum(dimensions) {
        return {
            namespace: 'AWS/ElastiCache',
            metricName: 'CacheHits',
            dimensions,
            statistic: 'Sum',
        };
    }
    static cacheMissesSum(dimensions) {
        return {
            namespace: 'AWS/ElastiCache',
            metricName: 'CacheMisses',
            dimensions,
            statistic: 'Sum',
        };
    }
    static replicationBytesSum(dimensions) {
        return {
            namespace: 'AWS/ElastiCache',
            metricName: 'ReplicationBytes',
            dimensions,
            statistic: 'Sum',
        };
    }
    static replicationLagAverage(dimensions) {
        return {
            namespace: 'AWS/ElastiCache',
            metricName: 'ReplicationLag',
            dimensions,
            statistic: 'Average',
        };
    }
    static bytesUsedForCacheSum(dimensions) {
        return {
            namespace: 'AWS/ElastiCache',
            metricName: 'BytesUsedForCache',
            dimensions,
            statistic: 'Sum',
        };
    }
    static currItemsSum(dimensions) {
        return {
            namespace: 'AWS/ElastiCache',
            metricName: 'CurrItems',
            dimensions,
            statistic: 'Sum',
        };
    }
    static casHitsSum(dimensions) {
        return {
            namespace: 'AWS/ElastiCache',
            metricName: 'CasHits',
            dimensions,
            statistic: 'Sum',
        };
    }
    static casMissesSum(dimensions) {
        return {
            namespace: 'AWS/ElastiCache',
            metricName: 'CasMisses',
            dimensions,
            statistic: 'Sum',
        };
    }
    static globalDatastoreReplicationLagAverage(dimensions) {
        return {
            namespace: 'AWS/ElastiCache',
            metricName: 'GlobalDatastoreReplicationLag',
            dimensions,
            statistic: 'Average',
        };
    }
}
exports.ElastiCacheMetrics = ElastiCacheMetrics;
