// @cloudcomponents/cdk-blue-green-container-deployment/lib/build-spec-generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BuildSpecGenerator = void 0;
class BuildSpecGenerator {
    constructor(spec) {
        this.spec = spec;
    }
    static empty() {
        return new BuildSpecGenerator({ version: '0.2' });
    }
    static default(props) {
        return new BuildSpecGenerator({
            version: '0.2',
            phases: {
                pre_build: {
                    commands: [
                        'echo Build started on `date`',
                        'echo Logging in to Amazon ECR...',
                        'aws --version',
                        `aws ecr get-login-password | docker login --username AWS --password-stdin ${props.account}.dkr.ecr.${props.region}.amazonaws.com`,
                        'COMMIT_HASH=$(echo $CODEBUILD_RESOLVED_SOURCE_VERSION | cut -c 1-7)',
                        'IMAGE_TAG=${COMMIT_HASH:=latest}',
                        'echo Replacing placeholder...',
                        'sed -i "s|SED_REPLACE_EXECUTION_ROLE_ARN|$EXECUTION_ROLE_ARN|g" taskdef.json',
                        'sed -i "s|SED_REPLACE_FAMILY|$FAMILY|g" taskdef.json',
                        'echo Current taskDefinition:',
                        'cat taskdef.json',
                    ],
                },
                build: {
                    commands: [
                        'echo Building the Docker image...',
                        'docker version',
                        'docker build -t $REPOSITORY_URI:latest -t $REPOSITORY_URI:$IMAGE_TAG -f Dockerfile .',
                    ],
                },
                post_build: {
                    commands: [
                        'echo Pushing the Docker images to container registry...',
                        'docker push $REPOSITORY_URI:latest',
                        'docker push $REPOSITORY_URI:$IMAGE_TAG',
                        'echo Writing image definitions file...',
                        'printf \'{"ImageURI":"%s"}\' $REPOSITORY_URI:$IMAGE_TAG > imageDetail.json',
                        'echo Build completed on `date`',
                    ],
                },
            },
            artifacts: {
                files: ['appspec.yaml', 'taskdef.json'],
                'secondary-artifacts': {
                    ManifestArtifact: {
                        files: ['appspec.yaml', 'taskdef.json'],
                    },
                    ImageArtifact: {
                        files: ['imageDetail.json'],
                    },
                },
            },
        });
    }
    render() {
        return this.spec;
    }
}
exports.BuildSpecGenerator = BuildSpecGenerator;
